### ISSUE TITLE
Phantomjs cannot find element where chrome and firefox can

### ISSUE THREAD

**Issue by aarkanov** [angular/protractor#585]
Hi,

I'm having an issue running protractor tests against phantomjs, and was wondering if I could get some help.

When I run the tests against firefox or chrome they pass no problem, however when I run the tests against phantomjs I get an error.

Here are all my steps:
1. start up webdriver with webdriver-manager start
2. the conf file looks as follows:

```
  exports.config = {

    seleniumAddress: 'http://localhost:4444/wd/hub',

    specs: [
    'PlayTests.js',
    ],

  multiCapabilities: [{
    'browserName': 'chrome'
  }, {
    'browserName': 'phantomjs'
  },{
    'browserName': 'firefox'
  }],



rootElement: 'body',

onPrepare: function () {
    // require('jasmine');
    require('jasmine-reporters');
    jasmine.getEnv().addReporter(
      new jasmine.JUnitXmlReporter('xmloutput', true, true));

        //     jasmine.getEnv().addReporter(new jasmine.JUnitXmlReporter(
    //         'outputdir/', true, true));
},

baseUrl: 'a base url',

jasmineNodeOpts: {
    // onComplete will be called just before the driver quits.
    onComplete: null,
    // If true, display spec names.
    isVerbose: false,
    // If true, print colors to the terminal.
    showColors: true,
    // If true, include stack traces in failures.
    includeStackTrace: true,
    // Default time to wait in ms before a test fails.
    defaultTimeoutInterval: 30000
  }
};
```
1. My test looks like this:

```
describe ('this is testing the new member website', function() {

    // var driver = null;
    beforeEach(function() {


        browser.get('/#');

    });

    it('will try to click on an element on the page', function() {
        element(by.linkText('News & Education')).click();
        var educationPage = element(by.tagName("body")).getText();
        expect(educationPage).toContain("Education");
    });

});

```
1. My error looks like this:

```

  1) this is testing the new member website will try to click on an element on the page
   Message:
     UnknownError: {"errorMessage":"Unable to find element with link text 'News & Education'","request":{"headers":{"Accept":"application/json, image/png","Connection":"Keep-Alive","Content-Length":"48","Content-Type":"application/json; charset=utf-8","Host":"localhost:17552"},"httpVersion":"1.1","method":"POST","post":"{\"using\":\"link text\",\"value\":\"News & Education\"}","url":"/element","urlParsed":{"anchor":"","query":"","file":"element","directory":"/","path":"/element","relative":"/element","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/element","queryKey":{},"chunks":["element"]},"urlOriginal":"/session/8d8fe850-a542-11e3-a78e-0bf0314fab61/element"}}
Build info: version: '2.39.0', revision: 'ff23eac', time: '2013-12-16 16:11:15'
System info: host: 'TestingMacs-Mac-mini.local', ip: '10.50.36.90', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.9.2', java.version: '1.6.0_65'
Driver info: driver.version: unknown
   Stacktrace:
     UnknownError: {"errorMessage":"Unable to find element with link text 'News & Education'","request":{"headers":{"Accept":"application/json, image/png","Connection":"Keep-Alive","Content-Length":"48","Content-Type":"application/json; charset=utf-8","Host":"localhost:17552"},"httpVersion":"1.1","method":"POST","post":"{\"using\":\"link text\",\"value\":\"News & Education\"}","url":"/element","urlParsed":{"anchor":"","query":"","file":"element","directory":"/","path":"/element","relative":"/element","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/element","queryKey":{},"chunks":["element"]},"urlOriginal":"/session/8d8fe850-a542-11e3-a78e-0bf0314fab61/element"}}
Build info: version: '2.39.0', revision: 'ff23eac', time: '2013-12-16 16:11:15'
System info: host: 'TestingMacs-Mac-mini.local', ip: '10.50.36.90', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.9.2', java.version: '1.6.0_65'
Driver info: driver.version: unknown
==== async task ====
WebDriver.findElement(By.linkText("News & Education"))
    at Protractor.findElement (/usr/local/lib/node_modules/protractor/lib/protractor.js:573:25)
    at Object.elementFinder.(anonymous function) [as click] (/usr/local/lib/node_modules/protractor/lib/protractor.js:88:24)
    at null.<anonymous> (/Users/TestingMac/WebEx/PlayTests.js:14:44)
    at /usr/local/lib/node_modules/protractor/jasminewd/index.js:54:12
    at wrapper [as _onTimeout] (timers.js:252:14)
==== async task ====
    at null.<anonymous> (/usr/local/lib/node_modules/protractor/jasminewd/index.js:53:12)
    at null.<anonymous> (/usr/local/lib/node_modules/protractor/node_modules/minijasminenode/lib/async-callback.js:45:37)
    at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
```

Any suggestions would be really appreciated.

**Comment by jcsmesquita** [comment]
Thanks @mcalthrop!

**Comment by mcalthrop** [comment]
de nada! :-)

**Comment by serpulga** [comment]
In config:

```javascript
onPrepare: function() {
        browser.driver.manage().window().setSize(1280, 1024);
    },
```

**Comment by mightymuke** [comment]
@mcalthrop - I could kiss you - thank you so much!

(Github needs a thanks button)

**Comment by glennypagaduan** [comment]
@mcalthrop Hi,

I am facing the same problem with and tried your suggestion. But seems not working for me. 

Issue: No element found using locator: by.model....


Here is my config file:
exports.config = {

	'seleniumAddress' : 'http://localhost:4444/wd/hub',

	'baseUrl' : 'http://....', <--- pointing in my local server to check my web app

	'capabilities' : {
		'browserName'                    : 'phantomjs',
		'phantomjs.binary.path'          : '/usr/local/bin/phantomjs',
		'phantomjs.ghostdriver.cli.args' : [ '--loglevel=DEBUG' ]
	},

	'allScriptsTimeout' : 100000,

	'getPageTimeout' : 100000,

	'specs' : [ 'test/e2e/login/login.spec.js' ],

	'framework'	: 'jasmine2',

	'jasmineNodeOpts' : {
		'showColors'             : true, 
		'defaultTimeoutInterval' : 100000,
		'showTiming'             : true,
		'print'                  : function () {},
		'realtimeFailure'        : true
	},

	'onPrepare' : function () {
		var SpecReporter = require( 'jasmine-spec-reporter' );
		browser.ignoreSynchronization = false;
		browser.driver.manage().window().setSize( 1280, 1024 );
		// browser.driver.manage().window();
		jasmine.getEnv().addReporter( new SpecReporter( { displayStacktrace : true } ) );
	}

};


Versions:
Protractor 2.0
Phantomjs 2.0
Jasmin 2

**Comment by dotjose** [comment]
My cases normally relates with this issue but mine is a little bit different. All my tests pass in firefox browser but when I change my browser config to phantomjs. All the tests failed. Every test suit failed why this is so? Is there any one who could help me?

**Comment by sallojusuresh** [comment]
https://github.com/angular/protractor/blob/master/docs/browser-support.md

**Comment by istinspring** [comment]
http://stackoverflow.com/questions/29632810/firefox-sees-element-where-phantomjs-does-not-when-using-selenium-webdriver

this one fixed issue for me

**Comment by michaelezehi** [comment]
Yep! I had similar issues and the sizing of PhantomJS was the culprit. Thanks for the heads up.

**Comment by skarensmoll** [comment]
Thank you ! @mcalthrop Your solution worked !
I really appreciate your help pretty much.

**Comment by capitolkrs** [comment]
`getDriver().manage().window().setSize(new Dimension(1280, 1024));`
if you are using java

**Comment by Shyam-Chen** [comment]
```js
// protractor.conf.js
const SpecReporter = require('jasmine-spec-reporter');
const phantomjs = require('phantomjs-prebuilt');
const babel = require('babel-core/register');

exports.config = {
  specs: [
    './src/**/*.e2e-spec.js'
  ],
  exclude: [],
  capabilities: {
    'browserName': 'phantomjs',
    'phantomjs.binary.path': phantomjs.path,
    'phantomjs.ghostdriver.cli.args': ['--loglevel=DEBUG']
  },
  onPrepare() {
    babel({ presets: ['latest'] });
    jasmine.getEnv().addReporter(new SpecReporter({ displayStacktrace: true }));
    global.webdriver = browser.driver;
    webdriver.ignoreSynchronization = true;
    webdriver.manage().window().setSize(1280, 768);
  },
  framework: 'jasmine',
  jasmineNodeOpts: {
    isVerbose: false,
    showColors: true,
    includeStackTrace: false
  }
};
```
```js
describe('About', () => {

  beforeEach(() => {
    webdriver.get('http://localhost:9876/about');
  });

  it('getCurrentUrl toMatch', () => {
    expect(webdriver.getCurrentUrl()).toMatch('/about');
  });

  it('getText toEqual', () => {
    let el = webdriver.findElement(by.css('#page p:nth-child(1)'));  // Unable to find element with css selector
    expect(el.getText()).toEqual('About');
  });

});
``

**Comment by edgarechm** [comment]
I have seen a lot commenting that this:

`browser.driver.manage().window().setSize(1280, 1024);`

...solved their problems. I tried and yes, solved the issue in my code. But I have not seen that anyone say WHY this solves the issue. WHY sticking to a window size of **1280x1024 pixels**?

If I use **_1 million by 1 millon pixels, problem solved FOREVER_**.

Why not? Why yes? @mcalthrop ?

**Comment by mcalthrop** [comment]
@edgarechm: the answer is in [this comment](#issuecomment-38064647). It relates to Twitter Bootstrap.

**Comment by AdamWital** [comment]
It still doesnt work for me..

**Comment by smajazayeri** [comment]
The same :( The error I receive is as follows:
```
Failed: {"errorMessage":"Unable to find element with css selector '*[id=\"cred_userid_inputtext\"]'","request":{"headers":{"Accept-Encoding":"gzip,deflate","Connection":"Keep-Alive","Content-Length":"66","Content-Type":"application/json; charset=utf-8","Host":"localhost:28755","User-Agent":"Apache-HttpClient/4.5.1 (Java/1.8.0_121)"},"httpVersion":"1.1","method":"POST","post":"{\"using\":\"css selector\",\"value\":\"*[id=\\\"cred_userid_inputtext\\\"]\"}","url":"/element","urlParsed":{"anchor":"","query":"","file":"element","directory":"/","path":"/element","relative":"/element","port":"","host":"","password":"","user":"","userInfo":"","authority":"","protocol":"","source":"/element","queryKey":{},"chunks":["element"]},"urlOriginal":"/session/dc848950-141d-11e7-b1ec-2fb7a4bbff83/element"}}
    For documentation on this error, please visit: http://seleniumhq.org/exceptions/no_such_element.html
    Build info: version: '2.53.1', revision: 'a36b8b1', time: '2016-06-30 17:37:03'
    System info: host: 'MININT-S1C4UPR', ip: '10.161.177.65', os.name: 'Windows 10', os.arch: 'x86', os.version: '10.0', java.version: '1.8.0_121'
    Driver info: driver.version: unknown
```

**Comment by pmros** [comment]
@AdamWital @smajazayeri maybe some javascript code is failing. Check phantomjs console log: `browser.log('browser')`. I had similar issues with pantomjs. Finally I solved it by adding polyfills (core.js), needed in phantomjs but not in chrome (a more modern browser).

**Comment by Shyam-Chen** [comment]
Switch to Chrome headless.

**Comment by edgarechm** [comment]
@pmros 
1: "browser.log is not a function." (sic) Message returned by protractor.
2:  What do you mean with "adding polyfills, needed in phantomjs"(sic)? how do you add "polyfills" to phantomjs?

@Shyam-Chen headless chrome is not an option in Windows. That only works on linux. But if you know how to do it for Windows, please tell us.

**Comment by pmros** [comment]
@edgarechm 
1. I used `browser.log` in [webdriverio](http://webdriver.io/api/protocol/log.html). I guess protactor has something similar.
2. I added polyfills to my html file, not to phantonjs specifically:
```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/core-js/2.4.1/core.min.js"></script>
```
I hope my comments can help someone.

**Comment by inconceivable** [comment]
In my case, the problem was that the class name of the element that I was looking for changed when I switched from Google Chrome to PhantomJS. I found this out by printing the page source from each browser, and piping the output to a file, where I could inspect it from the browser tools. The solution for me was that while the class names changed, the role attribute did not, so I could find the elements by xpath, where the role attribute was equal to the  constant role that did not change between browsers.
